main/lsp/request_dispatch.cc

```diff
@@ -21,8 +25,262 @@ class SorbetFenceTask final : public LSPTask {
+                    return false;
+                }
+            }
+        }
+    }
+    logger.debug("Taking fast path");
+    return true;
+}
+
+UnorderedMap<int, core::FileHash> mergeEvictions(const UnorderedMap<int, core::FileHash> &olderEvictions,
+                                                 const UnorderedMap<int, core::FileHash> &newerEvictions) {
+    // For evictions, which are needed for emulating an older `globalStateHashes`, we keep the oldest.
+    UnorderedMap<int, core::FileHash> combinedEvictions = olderEvictions;
+    for (auto &e : newerEvictions) {
+        if (!combinedEvictions.contains(e.first)) {
+            combinedEvictions[e.first] = e.second;
+        }
+    }
+    return combinedEvictions;
+}
+
 } // namespace
 
+LSPFileUpdates LSPLoop::commitEdit(SorbetWorkspaceEditParams &edit) {
+    LSPFileUpdates update;
+    update.epoch = edit.epoch;
+    update.editCount = edit.mergeCount + 1;
+    update.updatedFileHashes = LSPTypechecker::computeFileHashes(*config, edit.updates, *emptyWorkers);
+    update.updatedFiles = move(edit.updates);
+    update.canTakeFastPath = canTakeFastPath(*initialGS, *config, globalStateHashes, update);
+    update.cancellationExpected = edit.sorbetCancellationExpected;
+
+    // Update globalStateHashes. Keep track of file IDs for these files, along with old hashes for these files.
+    vector<core::FileRef> frefs;
+    UnorderedMap<int, core::FileHash> evictedHashes;
+    {
+        ENFORCE(update.updatedFiles.size() == update.updatedFileHashes.size());
+        core::UnfreezeFileTable fileTableAccess(*initialGS);
+        int i = -1;
+        for (auto &file : update.updatedFiles) {
+            auto fref = initialGS->findFileByPath(file->path());
+            i++;
+            if (fref.exists()) {
+                ENFORCE(fref.id() < globalStateHashes.size());
+                initialGS = core::GlobalState::replaceFile(move(initialGS), fref, file);
+            } else {
+                // This file update adds a new file to GlobalState.
+                update.hasNewFiles = true;
+                fref = initialGS->enterFile(file);
+                fref.data(*initialGS).strictLevel = pipeline::decideStrictLevel(*initialGS, fref, config->opts);
+                if (fref.id() >= globalStateHashes.size()) {
+                    globalStateHashes.resize(fref.id() + 1);
+                }
+            }
+            evictedHashes[fref.id()] = move(globalStateHashes[fref.id()]);
+            globalStateHashes[fref.id()] = update.updatedFileHashes[i];
+            frefs.push_back(fref);
+        }
+    }
+
+    // Index changes in initialGS. pipeline::index sorts output by file id, but we need to reorder to match the order of
+    // other fields.
+    UnorderedMap<u2, int> fileToPos;
+    {
+        int i = -1;
+        for (auto fref : frefs) {
+            // We should have ensured before reaching here that there are no duplicates.
+            ENFORCE(!fileToPos.contains(fref.id()));
```

Comment:
Ah, I guess this answers my question from above.  :-)
Does this ENFORCE kill just this thread (in which case, do we need to worry about mutations above?), or does it kill the running Sorbet entirely (in which case, I suppose it doesn't matter where it goes?)

Reply:
It kills the thread, but only in debug/test builds. In release builds, all `ENFORCE` statements are stripped out -- so we can liberally sprinkle them around.

They are meant to check invariants that _should always be true_ in your code.

Reply:
(ENFORCE also has some smarts regarding the debugger. If it detects an attached debugger on a failed ENFORCE, it will pause execution for inspection.)